pkg/ffi2abi/ffi.go

```diff
@@ -0,0 +1,421 @@
+		Events:      make([]*fftypes.FFIEvent, len(abi.Events())),
+	}
+	i := 0
+	for _, f := range abi.Functions() {
+		method, err := convertABIFunctionToFFIMethod(ctx, f)
+		if err != nil {
+			return nil, err
+		}
+		ffi.Methods[i] = method
+		i++
+	}
+	i = 0
+	for _, f := range abi.Events() {
+		event, err := convertABIEventToFFIEvent(ctx, f)
+		if err != nil {
+			return nil, err
+		}
+		ffi.Events[i] = event
+		i++
+	}
+	return ffi, nil
+}
+
+func convertABIFunctionToFFIMethod(ctx context.Context, abiFunction *abi.Entry) (*fftypes.FFIMethod, error) {
+	params := make([]*fftypes.FFIParam, len(abiFunction.Inputs))
+	returns := make([]*fftypes.FFIParam, len(abiFunction.Outputs))
+	details := map[string]interface{}{}
+	for i, input := range abiFunction.Inputs {
+		typeComponent, err := input.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, err
+		}
+		schema := getSchemaForABIInput(ctx, typeComponent)
+		param := &fftypes.FFIParam{
+			Name:   input.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		params[i] = param
+	}
+	for i, output := range abiFunction.Outputs {
+		typeComponent, err := output.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, err
+		}
+		schema := getSchemaForABIInput(ctx, typeComponent)
+		param := &fftypes.FFIParam{
+			Name:   output.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		returns[i] = param
+	}
+	if abiFunction.StateMutability != "" {
+		details["stateMutability"] = string(abiFunction.StateMutability)
+	}
+	if abiFunction.Payable {
+		details["payable"] = true
+	}
+	if abiFunction.Constant {
+		details["constant"] = true
+	}
+	return &fftypes.FFIMethod{
+		Name:    abiFunction.Name,
+		Params:  params,
+		Returns: returns,
+		Details: details,
+	}, nil
+}
+
+func convertABIEventToFFIEvent(ctx context.Context, abiEvent *abi.Entry) (*fftypes.FFIEvent, error) {
+	params := make([]*fftypes.FFIParam, len(abiEvent.Inputs))
+	details := map[string]interface{}{}
+	for i, output := range abiEvent.Inputs {
+		typeComponent, err := output.TypeComponentTreeCtx(ctx)
+		if err != nil {
+			return nil, err
+		}
+		schema := getSchemaForABIInput(ctx, typeComponent)
+		param := &fftypes.FFIParam{
+			Name:   output.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		params[i] = param
+	}
+	if abiEvent.Anonymous {
+		details["anonymous"] = true
+	}
+	return &fftypes.FFIEvent{
+		FFIEventDefinition: fftypes.FFIEventDefinition{
+			Name:    abiEvent.Name,
+			Params:  params,
+			Details: details,
+		},
+	}, nil
+}
+
+func getSchemaForABIInput(ctx context.Context, typeComponent abi.TypeComponent) *Schema {
+	schema := &Schema{
+		Details: &paramDetails{
+			Type:         typeComponent.Parameter().Type,
+			InternalType: typeComponent.Parameter().InternalType,
+			Indexed:      typeComponent.Parameter().Indexed,
+		},
+	}
+	switch typeComponent.ComponentType() {
+	case abi.ElementaryComponent:
+		t := GetFFIType(typeComponent.ElementaryType().String())
```

Comment:
Not laying down in the road of a merge on this one, but I do not think the `GetFFIType` function should exist any longer.

Instead, we should be using (/enhancing) the [ElementaryType](https://pkg.go.dev/github.com/hyperledger/firefly-signer/pkg/abi#ElementaryTypeInfo). The whole reason for existence of that object is to tell you metadata about the type.

Reply:
Being specific, I think there should be a function on the `ElementaryType` interface, that summarizes which of these functions are used for encoding/decoding:
https://github.com/hyperledger/firefly-signer/blob/main/pkg/abi/typecomponents.go#L146-L147

Reply:
As I pushed my latest commit I had the same thought about GetFFIType and that it probably shouldn't exist anymore. I hadn't worked out what would replace it yet though.

Reply:
PR up here with a proposal-in-code for this one: https://github.com/hyperledger/firefly-signer/pull/11